File: streamflowSignatures/python/streamflow_signatures/runoff_ratios.py
"""
Runoff ratio signatures.

Calculates annual and seasonal runoff ratios (Q/P).
Requires precipitation (PPT) data.
"""

from typing import Dict, List, Optional
import numpy as np
import pandas as pd
from .stats import generate_stats
from .config import RUNOFF_MIN_ANNUAL_PPT, RUNOFF_MIN_SEASONAL_PPT


def analyze_Q_PPT_relationships(
    streamflow_data: pd.DataFrame,
    seasonal_flags: Optional[List[Dict]] = None,
    min_annual_ppt: float = RUNOFF_MIN_ANNUAL_PPT,
    min_seasonal_ppt: float = RUNOFF_MIN_SEASONAL_PPT,
    trend_completeness: Optional[float] = None,
    decade_completeness: Optional[float] = None,
    min_values_for_stats: Optional[int] = None,
    changepoint: Optional[dict] = None,
    collector: Optional[object] = None,
) -> Dict[str, float]:
    """
    Calculate runoff ratio (Q/P) trends.

    Runoff ratio is the fraction of precipitation that becomes streamflow.

    Parameters
    ----------
    streamflow_data : pd.DataFrame
        Daily streamflow and climate data with columns:
            - water_year: int, water year
            - Q: float, daily discharge in mm/day
            - PPT: float, daily precipitation in mm/day
            - month: int, calendar month (1-12)
    min_annual_ppt : float, default 10
        Minimum annual precipitation (mm) for valid ratio
    min_seasonal_ppt : float, default 1
        Minimum seasonal precipitation (mm) for valid ratio

    Returns
    -------
    dict
        Dictionary of signature statistics with keys:
            - annual_runoff_ratio_*: Annual Q/P statistics
            - winter_runoff_ratio_*: Winter (Dec-Feb) Q/P statistics
            - spring_runoff_ratio_*: Spring (Mar-May) Q/P statistics
            - summer_runoff_ratio_*: Summer (Jun-Aug) Q/P statistics
            - fall_runoff_ratio_*: Fall (Sep-Nov) Q/P statistics

    Notes
    -----
    - Ratios >1 indicate more runoff than precipitation (e.g., snowmelt)
    - Ratios <1 indicate losses to ET, infiltration, etc.
    - Years/seasons with insufficient precipitation return NA
    """
    # Validate required columns
    required_cols = ["water_year", "Q", "PPT", "month"]
    missing = [c for c in required_cols if c not in streamflow_data.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")

    df = streamflow_data

    # Drop rows where either Q or PPT is NaN before summing
    # Matches R's aggregate(cbind(Q, PPT) ~ water_year, na.action=na.omit)
    df_valid = df.dropna(subset=["Q", "PPT"])

    # Calculate annual totals
    annual_totals = df_valid.groupby("water_year").agg({
        "Q": "sum",
        "PPT": "sum"
    }).reset_index()

    # Calculate annual runoff ratio with minimum PPT threshold
    annual_totals["annual_runoff_ratio"] = np.where(
        annual_totals["PPT"] > min_annual_ppt,
        annual_totals["Q"] / annual_totals["PPT"],
        np.nan
    )

    # Calculate seasonal totals and ratios
    def calc_seasonal_ratio(months, season_name):
        """Calculate seasonal runoff ratio."""
        seasonal = df_valid[df_valid["month"].isin(months)].copy()
        if len(seasonal) == 0:
            return pd.DataFrame({
                "water_year": annual_totals["water_year"],
                f"{season_name}_runoff_ratio": np.nan
            })

        totals = seasonal.groupby("water_year").agg({
            "Q": "sum",
            "PPT": "sum"
        }).reset_index()

        totals[f"{season_name}_runoff_ratio"] = np.where(
            totals["PPT"] > min_seasonal_ppt,
            totals["Q"] / totals["PPT"],
            np.nan
        )

        return totals[["water_year", f"{season_name}_runoff_ratio"]]

    # Calculate seasonal ratios
    winter_ratios = calc_seasonal_ratio([12, 1, 2], "winter")
    spring_ratios = calc_seasonal_ratio([3, 4, 5], "spring")
    summer_ratios = calc_seasonal_ratio([6, 7, 8], "summer")
    fall_ratios = calc_seasonal_ratio([9, 10, 11], "fall")

    # Merge all ratios
    all_ratios = annual_totals[["water_year", "annual_runoff_ratio"]].copy()

    for seasonal_df in [winter_ratios, spring_ratios, summer_ratios, fall_ratios]:
        all_ratios = all_ratios.merge(seasonal_df, on="water_year", how="left")

    # Apply seasonal completeness flags: set incomplete seasons to NA
    if seasonal_flags is not None and len(seasonal_flags) > 0:
        ratio_season_map = {
            "winter_runoff_ratio": "winter_complete",
            "spring_runoff_ratio": "spring_complete",
            "summer_runoff_ratio": "summer_complete",
            "fall_runoff_ratio": "fall_complete",
        }
        flags_df = pd.DataFrame(seasonal_flags)
        for ratio_col, flag_col in ratio_season_map.items():
            if ratio_col in all_ratios.columns and flag_col in flags_df.columns:
                merged = all_ratios[["water_year"]].merge(
                    flags_df[["water_year", flag_col]], on="water_year", how="left"
                )
                incomplete_mask = merged[flag_col].eq(False)
                all_ratios.loc[incomplete_mask, ratio_col] = np.nan

    # Get metric columns
    metric_cols = [c for c in all_ratios.columns if c != "water_year"]

    # Generate statistics
    result = generate_stats(
        all_ratios,
        value_cols=metric_cols,
        year_col="water_year",
        trend_completeness=trend_completeness,
        decade_completeness=decade_completeness, min_values_for_stats=min_values_for_stats, changepoint=changepoint, collector=collector,
    )

    # Count years where annual runoff ratio > 2.0 (per-gage scalar diagnostic)
    arr = all_ratios["annual_runoff_ratio"].values
    n_high = int(np.nansum(arr > 2.0))
    result["runoff_ratio_high_count"] = float(n_high)

    return result

File: streamflowSignatures/rpkg/R/runoff_ratios.R
#' Analyze Q-PPT relationships (runoff ratios)
#'
#' Computes annual and seasonal runoff ratios (Q/P) for each water year,
#' then applies generate_stats().
#'
#' @param streamflow_data A data.frame with columns: water_year, Q, PPT, month.
#' @param seasonal_flags Optional data.table from \code{preprocess_daily_data()}
#'   with per-year seasonal completeness flags. When supplied, seasonal runoff
#'   ratios for incomplete seasons are set to NA.
#' @return Named list of 8 statistics per ratio (5 ratios = 40 values).
#' @export
analyze_Q_PPT_relationships <- function(streamflow_data, seasonal_flags = NULL,
                                       trend_completeness = NULL,
                                       decade_completeness = NULL, min_values_for_stats = NULL, changepoint = NULL, collector = NULL) {
  required_cols <- c("water_year", "Q", "PPT", "month")
  missing <- setdiff(required_cols, colnames(streamflow_data))
  if (length(missing) > 0) stop(paste("Missing required columns:", paste(missing, collapse = ", ")))

  min_annual_ppt   <- pkg_env$runoff_min_annual_ppt
  min_seasonal_ppt <- pkg_env$runoff_min_seasonal_ppt

  # Drop rows where either Q or PPT is NA (paired masking, matches canonical R na.action=na.omit)
  complete_data <- streamflow_data[!is.na(streamflow_data$Q) & !is.na(streamflow_data$PPT), ]

  # Annual totals
  annual <- aggregate(cbind(Q, PPT) ~ water_year, data = complete_data, FUN = sum)
  annual$annual_runoff_ratio <- ifelse(annual$PPT > min_annual_ppt, annual$Q / annual$PPT, NA_real_)

  # Seasonal
  season_defs <- list(
    winter = c(12, 1, 2), spring = c(3, 4, 5),
    summer = c(6, 7, 8), fall = c(9, 10, 11)
  )

  all_ratios <- data.frame(water_year = annual$water_year)
  all_ratios <- merge(all_ratios, annual[, c("water_year", "annual_runoff_ratio")],
                      by = "water_year", all.x = TRUE)

  for (nm in names(season_defs)) {
    sub <- complete_data[complete_data$month %in% season_defs[[nm]], ]
    if (nrow(sub) > 0) {
      seas <- aggregate(cbind(Q, PPT) ~ water_year, data = sub, FUN = sum)
      col_name <- paste0(nm, "_runoff_ratio")
      seas[[col_name]] <- ifelse(seas$PPT > min_seasonal_ppt, seas$Q / seas$PPT, NA_real_)
      all_ratios <- merge(all_ratios, seas[, c("water_year", col_name)],
                          by = "water_year", all.x = TRUE)
    }
  }

  # Apply seasonal_flags: set incomplete seasons to NA
  if (!is.null(seasonal_flags)) {
    sf <- data.table::as.data.table(seasonal_flags)
    season_col_map <- list(
      winter_runoff_ratio = "winter_complete",
      spring_runoff_ratio = "spring_complete",
      summer_runoff_ratio = "summer_complete",
      fall_runoff_ratio   = "fall_complete"
    )
    for (metric_col in names(season_col_map)) {
      flag_col <- season_col_map[[metric_col]]
      if (metric_col %in% names(all_ratios) && flag_col %in% names(sf)) {
        for (i in seq_len(nrow(all_ratios))) {
          wy <- all_ratios$water_year[i]
          flag_row <- sf[sf$water_year == wy, ]
          if (nrow(flag_row) == 1 && !isTRUE(flag_row[[flag_col]])) {
            all_ratios[[metric_col]][i] <- NA_real_
          }
        }
      }
    }
  }

  metric_cols <- setdiff(names(all_ratios), "water_year")
  result <- generate_stats(all_ratios, value_cols = metric_cols, year_col = "water_year",
                           trend_completeness = trend_completeness,
                           decade_completeness = decade_completeness, min_values_for_stats = min_values_for_stats, changepoint = changepoint, collector = collector)

  # Count years where annual runoff ratio > 2.0 (per-gage scalar diagnostic)
  n_high <- sum(all_ratios$annual_runoff_ratio > 2.0, na.rm = TRUE)
  result$runoff_ratio_high_count <- as.numeric(n_high)

  result
}
